All changes for LoyaltyLion

clarkdave opened the pull request.

clarkdave committed:
tweak package.json, etc for our fork

clarkdave committed:
add support to disable connection pooling for postgres driver

clarkdave committed:
deprecate setLock in favour of lockMode

https://github.com/typeorm/typeorm/issues/2037

clarkdave committed:
add `lock` method to SelectQueryBuilder

https://github.com/typeorm/typeorm/issues/2037

clarkdave committed:
publish 0.2.9-1

tobenna commented on src/driver/postgres/PostgresDriver.ts:
```diff
@@ -657,7 +667,22 @@ export class PostgresDriver implements Driver {
+            };
+            const client = new this.postgres.Client(options);
+            await client.connect();
+            return [client, () => client.end()];
```
could we use `connection` here instead of client?

tobenna commented on tsconfig.json:
```diff
@@ -3,7 +3,7 @@
         "lib": ["es5", "es6"],
         "outDir": "build/compiled",
-        "target": "es5",
+        "target": "es2018",
```
🎉

tobenna reviewed (approved):
🏅

clarkdave committed:
publish 0.2.10-1

clarkdave committed:
add TransactionalEntityManager type

This type is used instead of a regular EntityManager in the
EntityManager.prototype.transaction callback. It extends an
EntityManager, so can be used as normal, but is not assignable to one,
so functions and services can require this type in order to assert they
receive a manager that has already started a transaction

clarkdave committed:
publish 0.2.10-2

clarkdave committed:
publish 0.2.10-3

atreidesend committed:
set and unset activeTransaction flag on databaseConnection

LL specific change - our legacy services use this field to determine
if a transaction is active or not, if we don't manage it via
typeorm then it's possible legacy clients will still think they
are in a transaction when they're not

atreidesend committed:
update package to 0.2.10-4

clarkdave committed:
use esnext compiler target and add ts version to vscode settings

clarkdave committed:
remove bad connections from the postgres pool after a failed query

typeorm currently lets pg.pool assume that clients are always safe to
add back into the pool, even if it encountered an error while running
a query. this is /usually/ fine, because the vast majority of errors
that happen during a query do not cause the client's connection itself
to become inoperable

however, if, during a query, the client's connection is killed, the
query will fail and the client will be placed back into the pool even
though it's linked to a now broken connection, i.e. that client will
never work again, but because pg.Pool doesn't know that, it'll keep
handing it out. end result: lots of failed queries

this patches `Connection.prototype.query` so that it passes the error
back to the driver when it calls `queryRunner.release`. the driver can
(if applicable) then inspect the error and decide if it should result
in the client being removed from the pool or just released as normal

clarkdave committed:
publish 0.2.20-1

clarkdave committed:
remove clients from the pool following any error

we're not confident in our ability to accurately determine what a "bad"
error (i.e. one that has rendered the connection broken). to be safe,
we'll instruct the pool to remove any client that failed to execute
a query

though this will result in more pool churn, we're using pgbouncer so
those connections are cheap, and this provides maximum safety overall

Commit:
Bumped node version to 13.7.0

Commit:
Bump to version 0.2.20-3

draaglom committed:
test: add benchmark for select querybuilder (#8955)

Motivation: the query builder (and within it, replacePropertyNames and
associated functions) is pretty CPU intensive. For our workload, it's
one of the hottest functions in our entire stack.

While improved in https://github.com/typeorm/typeorm/pull/4760,
There are still outstanding issues relating to perf e.g. https://github.com/typeorm/typeorm/issues/3857

As we all know though, the first step in optimization is to measure
systematically ;)

https://wiki.c2.com/?ProfileBeforeOptimizing

On my machine, this benchmark runs in ~3500ms or about 0.35ms/query.
This tells us there's a way to go - on my stack, that's about 1/3 of a
typical query's latency!

draaglom committed:
perf: don't recompile escapeRegExp for every query (#8956)

Context: the query builder is pretty CPU intensive, and can be slow -
e.g. https://github.com/typeorm/typeorm/issues/3857

One of the things which makes this slow is `escapeRegExp` in the query
builder: we freshly construct the same RegExp once per
`replacePropertyName` invocation (many times per overall query!) and
since the RegExp itself is constant -- we can lift it out and construct
it once.

Over-all this saves about 8% on our query build times as measured by
 #8955.

draaglom committed:
perf: partially lift matching from regexp to js (#9032)

Digging further into #3857.

See also #8955, #8956.

As [previously
discussed](https://github.com/typeorm/typeorm/issues/3857#issuecomment-699505893),
the query builder currently suffers from poor performance in two ways:
quadratic numbers of operations with respect to total table/column
counts, and poor constant factor performance (regexps can be expensive
to build/run!)

The constant-factor performance is the more tractable problem: no longer
quadratically looping would be a chunky rewrite of the query builder,
but we can locally refactor to be a bunch cheaper in terms of regexp
operations.

This change cuts the benchmark time here in ~half (yay!).

We achieve this by simplifying the overall replacement regexp (we don't
need our column names in there, since we already have a plain object
where they're the keys to match against) so compilation of that is much
cheaper, plus skipping the need to `escapeRegExp` every column as a
result.

frangz committed:
Add instructions to build on ARM

frangz committed:
Increment version

frangz committed:
Fix multiple-joins-querybuilder test
